sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/identifiers.scala

```diff
@@ -29,12 +28,23 @@ sealed trait IdentifierWithDatabase {
 
   def database: Option[String]
 
+  /*
+   * Escapes back-ticks within the identifier name with double-back-ticks.
+   */
+  private def quoteIdentifier(name: String): String = name.replace("`", "``")
+
   def quotedString: String = {
-    if (database.isDefined) s"`${database.get}`.`$identifier`" else s"`$identifier`"
+    val replacedId = quoteIdentifier(identifier)
+    val replacedDb = database.map(quoteIdentifier(_))
+
+    if (replacedDb.isDefined) s"`${replacedDb.get}`.`$replacedId`" else s"`$replacedId`"
   }
 
   def unquotedString: String = {
-    if (database.isDefined) s"${database.get}.$identifier" else identifier
```

Comment:
I think we should keep `unquotedString ` this as it is. That gives some a more readable (less parseable) version of the identifier. I just added an example with `unquotedIdentifier` to show that there was no way to get a parseable string from the table identifier.